### DISCUSSION TITLE
Unable to connect to public IP address through VM

### DISCUSSION THREAD

**Discussion by konsweb-122** [EY-Data-Science-Program/2021-Better-Working-World-Data-Challenge#55] [3.0 VMs and data (Azure VM, local deployment, and Jupyter)]
I have encountered the following problem of not being able to access to the public IP address after deploying my VM: 
<img width="319" alt="Picture 1" src="https://user-images.githubusercontent.com/83834706/117525404-b5170d80-aff4-11eb-9031-f57354276239.png">

I have run the code according to 4a. Troubleshooting FAQ. The state of odc_jupyter_1 is "Restarting", how do i overcome it? 
<img width="452" alt="Picture2" src="https://user-images.githubusercontent.com/83834706/117525399-b21c1d00-aff4-11eb-9e8f-16c9d8671e50.png">

Or are there other steps that I should do? Thanks

**Comment by EY-glowglowglow** [comment]
@codeindulgence  any ideas on this?

**Comment by nikunj-taneja** [comment]
I am facing the same problem! Has anyone found a way to resolve it yet? For reference, I have attached a log file that I retrieved after accessing the VM Terminal through ssh. 
[log.txt](https://github.com/EY-Data-Science-Program/2021-Better-Working-World-Data-Challenge/files/6470682/log.txt)

**Reply by konsweb-122** [reply]
Btw, how do you get the log file? I can only see in the command prompt?

**Reply by nickbutler** [reply]
Hey @konsweb-122, take a look at [Check the Install Log](https://github.com/EY-Data-Science-Program/2021-Better-Working-World-Data-Challenge/wiki/4a.-Troubleshooting:-FAQ#check-the-install-log) from the [FAQ](https://github.com/EY-Data-Science-Program/2021-Better-Working-World-Data-Challenge/wiki/4a.-Troubleshooting:-FAQ)

**Reply by konsweb-122** [reply]
Thanks

**Comment by nickbutler** [comment]
Hi folks,

So there seems to be an issue with the Jupyter container causing it to be stuck in "Restarting".
Let's see if we can find out why by running the below from `/opt/odc/`:

```
sudo docker-compose logs --tail=all --no-color jupyter | tee jupyter.log
```

This will print the container logs and also write them to a file `/opt/odc/jupyter.log`.

This may provide some insight into what is causing the "Restarting" state.

Depending on what that reveals (if anything), we can also try manually restarting the container again and seeing if that has any effect:

```
sudo docker-compose restart
```

**Reply by konsweb-122** [reply]
I got this error: 

jupyter_1   | [C 12:52:27.847 NotebookApp] Bad config encountered during initialization:
jupyter_1   | [C 12:52:27.847 NotebookApp] The 'token' trait of a NotebookApp instance must be a unicode string, but a value of 1234 <class 'int'> was specified.

May I know how can i solve it? I tried sudo docker-compose restart as well but it didnt work as well

**Reply by nickbutler** [reply]
Oh, that's interesting! So it looks like the Jupyter password you have chosen is all numerical and is interpreted as an integer when Jupyter loads it.

To fix this please use a password which includes alphabetical characters, also. 

You'll need to recreate the containers:

```
cd /opt/odc
sudo docker-compose down
sudo ./install-cube.sh MYALPHANUMERICPASSWORD
```

**Reply by konsweb-122** [reply]
Thanks a lot, @codeindulgence!
I have changed my password and run sudo docker-compose logs --tail=all --no-color jupyter | tee jupyter.logthis
But this error occurred:

jupyter_1 | [W 06:29:24.732 NotebookApp] No web browser found: could not locate runnable browser.

May I know how to resolve this?

**Reply by nickbutler** [reply]
Hi @konsweb-122,
That warning is to be expected, Jupyter is trying to open a browser, but since it's running in server environment there isn't one. Normally if you're running this on your laptop/desktop it'll automatically open a web browser to Jupyter.

If there's no other error, and the containers are running, you should be able to load Jupyter by going to the VM IP in your browser!

**Reply by konsweb-122** [reply]
Thanks. Everything is fine now. But I still cant access to the jupter notebook after copying the link. The browser displayed: 

"This site can’t be reached localhost refused to connect."

Is there anything else I should do?

**Reply by EY-glowglowglow** [reply]
@codeindulgence can you please have a look at this? thanks!

**Reply by nickbutler** [reply]
Hi @konsweb-122, where do you see this error? 
If you're connecting to Jupyter from your local machine browser, you should use the VM IP.

Other things you can check:
* Confirm the status of the container (using `sudo docker-compose ps` as previously)
* Confirm the "ports" value of the Jupyter container in the above command
* Check the VM host can access Jupyter: `curl localhost` - should print the Jupyter HTML
* Test access using port 8888 by adding `:8888` to the VM IP in your browser address bar.

Cheers

**Reply by konsweb-122** [reply]
This is my result for first 2 points 
ubuntu@CubeInABox:/opt/odc$ sudo docker-compose ps
     Name                   Command               State                    Ports
--------------------------------------------------------------------------------------------------
odc_jupyter_1    /bin/tini -s -- with_boots ...   Up      0.0.0.0:80->8888/tcp,:::80->8888/tcp
odc_postgres_1   /bin/sh -c /docker-entrypo ...   Up      0.0.0.0:5432->5432/tcp,:::5432->5432/tcp

**Reply by konsweb-122** [reply]
typed this in the command prompt, nothing happens: ubuntu@CubeInABox:~$ curl localhost

i also tried adding ipaddress:8888 (ex: 13.75.138.9:8888) in browser but nothing happens as well

**Reply by nickbutler** [reply]
I'm officially stumped on this one. Everything looks fine based on the output.

All I can suggest now is either:
* Deploy a new environment (and delete this one)
* Try a [local deployment](https://github.com/EY-Data-Science-Program/2021-Better-Working-World-Data-Challenge/wiki/1c.-Getting-started:-Local-Environment)

**Reply by konsweb-122** [reply]
Ok thank you @codeindulgence
